### PR TITLE
[Configuration] When a multiple-entry field is added duplication might occur

### PR DESCRIPTION
**Pull request by kongtiaowang**
This pull request fixes the bug when a multiple-entry field is added duplication might occur.
https://redmine.cbrain.mcgill.ca/issues/13808

### CODE REVIEW

File: modules/configuration/ajax/process.php [hunk 1]
```diff
@@ -29,7 +29,7 @@
 $client->initialize();
 
 $DB =& Database::singleton();
-foreach ($_POST as $key => $value) {
+foreach (array_unique($_POST) as $key => $value) {
     if (is_numeric($key)) { //update
```
Comment (PapillonMcGill) [inline modules/configuration/ajax/process.php]: By duplication, you mean duplicate value? 
or duplicate key?

This patch will prevent multiple Config settings having the same value, should check for same key=value pair.  
You could also add a unique constraint in the DB on (ConfigID, Value). This way any attempt to add a duplication would generate a DB error

Comment (driusan) [reply modules/configuration/ajax/process.php]: @xlecours @ridz1208 @MounaSafiHarab @christinerogers Can any of you think of any reasons why the same config option would need to have the same value specified multiple times with the current LORIS config options?

Comment (kongtiaowang) [reply modules/configuration/ajax/process.php]: @PapillonMcGill 
Thanks for your comment.  Could u review again? I have added a unique key to the table.

Comment (PapillonMcGill) [reply modules/configuration/ajax/process.php]: Nice to change TEXT to VARCHAR(768) but is the 768 value a little bit high? (maybe 100 or 150?)
With the UNIQUE key in DB, a DB error would be generated preventing multiple setting having the same value. 

But still using **array_unique**, that would keep only the first occurrence of a $value even if multiple  $key can have the same value (think YES/NO field)

Comment (kongtiaowang) [reply modules/configuration/ajax/process.php]: 100 or 150 are not enough to save the data. The value needs to save "HTML code" and Document. But you are right, "array_unique" is a good choice in here.

**Comment by kongtiaowang** [comment]
@PapillonMcGill Please review again.

File: SQL/Archive/19.0/2018-02-20_add_unique_into_Config.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/*  delete duplicate ConfigID-value pairs */
```
Comment (ridz1208) [inline SQL/Archive/19.0/2018-02-20_add_unique_into_Config.sql]: this should not be going into 19.0 directory

File: SQL/0000-00-03-ConfigTables.sql [hunk 3]
```diff
@@ -29,15 +32,16 @@ CREATE TABLE `ConfigSettings` (
 CREATE TABLE `Config` (
   `ID` int(11) NOT NULL AUTO_INCREMENT,
   `ConfigID` int(11) NOT NULL,
-  `Value` text,
+  `Value` varchar(767),
```
Comment (ridz1208) [inline SQL/0000-00-03-ConfigTables.sql]: Since we do not set a limit to the size of text that we can enter, i think keeping TEXT is safer !! (especially that non-strict databases do not throw an error on save if data is truncated !)

this will also reduce the amount of changes in this PR

File: modules/configuration/js/configuration_helper.js [hunk 4]
```diff
@@ -86,6 +86,9 @@ $(function () {
             error: function(xhr, desc, err) {
                 console.log(xhr);
                 console.log("Details: " + desc + "\nError:" + err);
+                var html = "<label>Internal Server Error or Duplicate Values</label>";
```
Comment (ridz1208) [inline modules/configuration/js/configuration_helper.js]: We should remove the "Internal Server Error" and catch separate errors and display the message appropriately

@driusan ??

File: SQL/0000-00-03-ConfigTables.sql (hunk 3 above)
Comment (driusan) [reply SQL/0000-00-03-ConfigTables.sql]: I agree with @ridz1208 .. there's no reason for any changes other than the check constraint (and patch) to accomplish what the title/description says the problem is.

File: modules/configuration/js/configuration_helper.js (hunk 4 above)
Comment (driusan) [reply modules/configuration/js/configuration_helper.js]: yeah, there's already details going to the console if someone needs to figure out what happened. I think we can improve the UI in a different PR if necessary.. but errors should be exceptional (I hope?), not a normal occurance, so i'm not sure how often this code path will even be triggered.

**Comment by ridz1208** [comment]
@kongtiaowang after the discussions w've had on this what I think should be done is

- keep the sql in the Patch to clean-up surrent duplicated data
- make a php check before the processing making sure that the values submitted do not already exist in the database
- make sure that if the submit button is clicked multiple times in a row by accident, the values dont get added submitted times (currently the case)

Here we are trying to simulate a unique key in PHP because adding a unique key to SQL is problematic

File: modules/configuration/ajax/process.php [hunk 5]
```diff
@@ -27,7 +27,24 @@
 $client = new NDB_Client();
 $client->makeCommandLine();
 $client->initialize();
-
+// this qurey could delete duplicate ConfigID-value pairs 
+$query = "Delete from Config  Where ID in
```
Comment (PapillonMcGill) [inline modules/configuration/ajax/process.php]: // this qu**er**y could ...

File: modules/configuration/ajax/process.php [hunk 6]
```diff
@@ -39,6 +56,8 @@
                 array('Value' => $value),
                 array('ID' => $key)
             );
+            //delete duplicate ConfigID-value pairs 
+            $DB->run($query);
```
Comment (PapillonMcGill) [inline modules/configuration/ajax/process.php]: I would put this query execution after the insert

File: modules/configuration/ajax/process.php (hunk 5 above)
Comment (ridz1208) [reply modules/configuration/ajax/process.php]: the query should be entirely removed

File: modules/configuration/ajax/process.php (hunk 6 above)
Comment (ridz1208) [reply modules/configuration/ajax/process.php]: the query should be entirely removed

**Comment by kongtiaowang** [comment]
@ridz1208 It is done, please review.

File: modules/configuration/ajax/process.php [hunk 7]
```diff
@@ -27,39 +27,100 @@
 $client = new NDB_Client();
 $client->makeCommandLine();
 $client->initialize();
-
+// this qurey could delete duplicate ConfigID-value pairs
 $DB =& Database::singleton();
```
Comment (PapillonMcGill) [inline modules/configuration/ajax/process.php]: // this qu**er**y delete 

fix the type and remove "could"

**Comment by kongtiaowang** [comment]
replace by https://github.com/aces/Loris/pull/3567

**Comment by driusan** [comment]
@kongtiaowang why was this re-opened if it's replaced by 3567?

**Comment by kongtiaowang** [comment]
@driusan I rebased this PR to major branch.To check if this PR Could pass travis_CI, I will close the new one.

File: modules/configuration/js/configuration_helper.js [hunk 8]
```diff
@@ -86,6 +86,9 @@ $(function () {
             error: function(xhr, desc, err) {
                 console.log(xhr);
                 console.log("Details: " + desc + "\nError:" + err);
+                var html = "<label>Duplicate Values</label>";
```
Comment (zaliqarosli) [inline modules/configuration/js/configuration_helper.js]: Could maybe make the error more explicit i.e. 'One or more fields with multiple-entry option has duplicate values'

File: SQL/Archive/19.1/2018-02-20_remove_duplicate_value_from_Config.sql [hunk 9]
```diff
@@ -0,0 +1,10 @@
+/*  delete duplicate ConfigID-value pairs */
```
Comment (zaliqarosli) [inline SQL/Archive/19.1/2018-02-20_remove_duplicate_value_from_Config.sql]: should this patch maybe be put in one of the new directories, either `Cleanup_patches` or `New_patches`?

File: modules/configuration/ajax/process.php [hunk 10]
```diff
@@ -27,39 +27,91 @@
...
+ * @param string $key   The value of the key
+ * @param string $value The value of the value
+ *
+ * @return string $result
+ */
+function checkDuplicate($key,$value)
+{
+       $DB     =& Database::singleton();
+       $result = $DB->pselectOne(
+           "Select count(*) from Config where ConfigID =:ConfigID and Value =:Value",
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: Capitalize select statement to be consistent?

File: modules/configuration/ajax/process.php [hunk 11]
```diff
@@ -27,39 +27,91 @@
...
-
+/**
+ * Check Duplicate value
+ *
+ * @param string $key   The value of the key
+ * @param string $value The value of the value
+ *
+ * @return string $result
+ */
+function checkDuplicate($key,$value)
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: could maybe change name of this function to countDuplicate, and the other one duplicateExists, or something along the lines of being more descriptive

Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: could maybe change name of this function to countDuplicate, and the other one duplicateExists, or something along the lines of being more descriptive

File: SQL/Archive/19.1/2018-02-20_remove_duplicate_value_from_Config.sql (hunk 9 above)
Comment (kongtiaowang) [reply SQL/Archive/19.1/2018-02-20_remove_duplicate_value_from_Config.sql]: We don't have the folder as you said.
So, the best way is put it in the patch.

File: modules/configuration/ajax/process.php [hunk 12]
```diff
@@ -76,11 +76,11 @@
  *
  * @return string $result
  */
-function checkDuplicate($key,$value)
+function countDuplicate($key,$value)
 {
        $DB     =& Database::singleton();
        $result = $DB->pselectOne(
-           "Select count(*) from Config where ConfigID =:ConfigID and Value =:Value",
+           "SELECT count(*) FROM Config WHERE ConfigID =:ConfigID AND Value =:Value",
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: capitalize COUNT(*)?

**Comment by kongtiaowang** [comment]
@ridz1208 @driusan @PapillonMcGill @cmadjar Please review again.

**Comment by kongtiaowang** [comment]
@cmadjar Thank you for review, the error message is not good enough. I can open another PR to make the error message more friendly for the users. Like you said, a warning popup. So far, I just followed the old warning style.

**Comment by cmadjar** [comment]
> @cmadjar Thank you for the comment, the error message is not good enough. I can open another PR to make the error message more friendly for the users. Like you said, a warning popup. So far, I just followed the old warning style.

@kongtiaowang since you will issue another PR for the error message, approving this PR as it was my only change request.

File: SQL/Cleanup_patches/2018-02-20_remove_duplicate_value_from_Config.sql [hunk 13]
```diff
@@ -0,0 +1,10 @@
+/*  delete duplicate ConfigID-value pairs */
+
+Delete from Config  Where ID in
+    (
+      select temp.ID from
+       (select ID from Config c
+          where (c.ConfigId,c.Value) in (select ConfigId,Value from Config group by ConfigId,Value having count(*) > 1)
+          and ID not in (select min(ID) from Config group by ConfigId,Value having count(*)>1)
```
Comment (driusan) [inline SQL/Cleanup_patches/2018-02-20_remove_duplicate_value_from_Config.sql]: I don't think the logic of this is correct because of the number of sub-queries. The `MIN()` from the `ID NOT IN` part is referencing a different ID than the first part of the where clause. (At any rate, it's hard to follow enough to verify..)

Couldn't this be done with a single group by?

File: modules/configuration/ajax/process.php [hunk 14]
```diff
@@ -27,39 +27,91 @@
...
-            );
+            // if no duplicate value then do updating
+            if (duplicateExistsInDropdown($key, $value)) {
+                $DB->update(
+                    'Config',
+                    array('Value' => $value),
+                    array('ID' => $key)
+                );
+            } else {
+                   header("HTTP/1.1 303 Duplicate value for update");
```
Comment (driusan) [inline modules/configuration/ajax/process.php]: This isn't a valid HTTP response code. 303 is 'See Other', intended for redirects. 

I think it should just be a 400 Bad Request (or maybe a 409 Conflict) response.

File: modules/configuration/ajax/process.php [hunk 15]
```diff
@@ -27,39 +27,91 @@
...
 }
-
-
+/**
+ * Check Duplicate value
+ *
+ * @param string $key   The value of the key
+ * @param string $value The value of the value
+ *
+ * @return string $result
```
Comment (driusan) [inline modules/configuration/ajax/process.php]: Shouldn't the string be an int?

File: modules/configuration/ajax/process.php [hunk 16]
```diff
@@ -27,39 +27,91 @@
...
+ */
+function countDuplicate($key,$value)
+{
+       $DB     =& Database::singleton();
+       $result = $DB->pselectOne(
+           "SELECT count(*) FROM Config WHERE ConfigID =:ConfigID AND Value =:Value",
```
Comment (driusan) [inline modules/configuration/ajax/process.php]: `COUNT` should be capitalized, it's a built-in MySQL function.

File: modules/configuration/ajax/process.php [hunk 17]
```diff
@@ -27,39 +27,91 @@
...
+           array(':ID' => $id)
+       );
+       $IDBefore = $DB->pselectOne(
+           "SELECT ID FROM Config WHERE ConfigID =:ConfigID AND Value =:Value",
+           array(
+            ':ConfigID' => $ConfigID,
+            ':Value'    => $value,
+           )
+       );
+       return ($id == $IDBefore || $IDBefore == null);
```
Comment (driusan) [inline modules/configuration/ajax/process.php]: I don't think this is reliable. You seem to be relying on the order that MySQL is returning rows without an explicit `ORDER BY` clause.

File: SQL/Cleanup_patches/2018-02-20_remove_duplicate_value_from_Config.sql (hunk 13 above)
Comment (PapillonMcGill) [reply SQL/Cleanup_patches/2018-02-20_remove_duplicate_value_from_Config.sql]: @driusan The query is long and complex but it make sense to select all ID except the first one (lowest ID) in case of duplication.

File: SQL/Cleanup_patches/2018-02-20_remove_duplicate_value_from_Config.sql [hunk 18]
```diff
@@ -0,0 +1,10 @@
+/*  delete duplicate ConfigID-value pairs */
+
+Delete from Config  Where ID in
```
Comment (xlecours) [inline SQL/Cleanup_patches/2018-02-20_remove_duplicate_value_from_Config.sql]: No need for subqueries
`DELETE FROM Config USING Config, Config c1 WHERE Config.ID < c1.ID AND Config.ConfigID = c1.ConfigID AND Config.Value = c1.Value;`

Comment (xlecours) [reply SQL/Cleanup_patches/2018-02-20_remove_duplicate_value_from_Config.sql]: If you want to keep the olders then just change `Config.ID < c1.ID` for `Config.ID < c1.ID`

Comment (kongtiaowang) [reply SQL/Cleanup_patches/2018-02-20_remove_duplicate_value_from_Config.sql]: @xlecours Thank you, you made it more clearly for everyone.

File: modules/configuration/ajax/process.php (hunk 14 above)
Comment (kongtiaowang) [reply modules/configuration/ajax/process.php]: Thanks, done.

File: modules/configuration/ajax/process.php (hunk 17 above)
Comment (kongtiaowang) [reply modules/configuration/ajax/process.php]: After running the patch, all the ConfigID and Value pairs are unique, so no need to using the ORDER  BY in the query. I think.

File: modules/configuration/ajax/process.php (hunk 15 above)
Comment (kongtiaowang) [reply modules/configuration/ajax/process.php]: done

**Comment by kongtiaowang** [comment]
@driusan Please review again.

File: modules/configuration/ajax/process.php [hunk 19]
```diff
@@ -27,39 +27,100 @@
...
         if ($value == "") {
             $DB->delete('Config', array('ID' => $key));
         } else {
-            $DB->update(
-                'Config',
-                array('Value' => $value),
-                array('ID' => $key)
-            );
+            // if no duplicate value then do updating
+            if (duplicateExistsInDropdown($key, $value)) {
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: the comment and the name of the function is contradictory

File: modules/configuration/ajax/process.php [hunk 20]
```diff
@@ -27,39 +27,100 @@
...
+}
+/**
+ * Check dropdown list Duplicate value
+ *
+ * @param string $id    The value of the id
+ * @param string $value The value of the value
+ *
+ * @return boolean return true if there is no Duplicate value
+ */
+function duplicateExistsInDropdown($id,$value)
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: this is confusing. if the function returns true when duplicate does not exist, then the name of the function shouldn't be 'duplicateExists'

File: modules/configuration/ajax/process.php [hunk 21]
```diff
@@ -27,39 +27,100 @@
...
+{
+       $DB =& Database::singleton();
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: could we call \NDB_Factory->database() instead while we're at it? same goes for line 30 of `modules/configuration/ajax/process.php`

File: modules/configuration/ajax/process.php [hunk 22]
```diff
@@ -96,9 +98,10 @@ function countDuplicate($key,$value)
  *
  * @return boolean return true if there is no Duplicate value
  */
-function duplicateExistsInDropdown($id,$value)
+function duplicateNotExistsInDropdown($id,$value)
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: what about "'noDuplicateInDropdown"?

Comment (kongtiaowang) [reply modules/configuration/ajax/process.php]: OK, good Idea.

File: modules/configuration/ajax/process.php [hunk 23]
```diff
@@ -27,39 +27,103 @@
 $client = new NDB_Client();
 $client->makeCommandLine();
 $client->initialize();
-
-$DB =& Database::singleton();
+$factory = NDB_Factory::singleton();
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: sorry, hopefully one last thing! you want a \ before NDB_Factory. for the same lines below aswell

File: modules/configuration/js/configuration_helper.js [hunk 24]
```diff
@@ -86,6 +86,9 @@ $(function () {
             error: function(xhr, desc, err) {
                 console.log(xhr);
                 console.log("Details: " + desc + "\nError:" + err);
```
Comment (paiva) [inline modules/configuration/js/configuration_helper.js]: Do we need the console log prints if we are returning the error in html?

File: modules/configuration/ajax/process.php [hunk 25]
```diff
@@ -27,39 +27,103 @@
...
+ * Check Duplicate value
+ *
+ * @param string $key   The value of the key
+ * @param string $value The value of the value
+ *
+ * @return int $result
+ */
+function countDuplicate($key,$value)
+{
+    $factory   = NDB_Factory::singleton();
```
Comment (paiva) [inline modules/configuration/ajax/process.php]: I think it should be `\NDB_Factory::` as @zaliqarosli suggested

File: modules/configuration/ajax/process.php [hunk 26]
```diff
@@ -27,39 +27,103 @@
...
+ * Check dropdown list Duplicate value
+ *
+ * @param string $id    The value of the id
+ * @param string $value The value of the value
+ *
+ * @return boolean return true if there is no Duplicate value
+ */
+function noDuplicateInDropdown($id,$value)
+{
+    $factory = NDB_Factory::singleton();
```
Comment (paiva) [inline modules/configuration/ajax/process.php]: Here too `\NDB_Factory::`

File: modules/configuration/ajax/process.php [hunk 27]
```diff
@@ -27,39 +27,103 @@
...
+ * @return int $result
+ */
+function countDuplicate($key,$value)
+{
+    $factory   = \NDB_Factory::singleton();
+    $DB        = $factory->database();
+       $result = $DB->pselectOne(
+           "SELECT COUNT(*) FROM Config WHERE ConfigID =:ConfigID AND Value =:Value",
+           array(
+            ':ConfigID' => $key,
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: i don't think there should be a `:` before ConfigID, or Value

File: modules/configuration/ajax/process.php [hunk 28]
```diff
@@ -27,39 +27,103 @@
...
+       $ConfigID = $DB->pselectOne(
+           "SELECT ConfigID FROM Config WHERE ID =:ID",
+           array(':ID' => $id)
+       );
+       // IDBefore means that row ID contains the same configID and value pair.
+       $IDBefore = $DB->pselectOne(
+           "SELECT ID FROM Config WHERE ConfigID =:ConfigID AND Value =:Value",
+           array(
+            ':ConfigID' => $ConfigID,
+            ':Value'    => $value,
```
Comment (zaliqarosli) [inline modules/configuration/ajax/process.php]: same issue here about unnecessary extra `:` and 2 lines above

**Comment by kongtiaowang** [comment]
@driusan Could u review again?

**Comment by zaliqarosli** [comment]
^ forgot to run patch, tested and works!

File: SQL/Archive/19.0/2018-02-20_add_unique_into_Config.sql (hunk 2 above)
Comment (kongtiaowang) [reply SQL/Archive/19.0/2018-02-20_add_unique_into_Config.sql]: done

**Comment by zaliqarosli** [comment]
@driusan should your review have been automatically dismissed?
